### PR TITLE
[candidate] Create Sex table and remove Sex enum

### LINKED ISSUE #8936: Sex should not be an enum
Sex is currently defined in the candidate table as an enum with the values `enum('Male','Female','Other')`. This is to align with the possible DICOM values, but is archaic from a scientific perspective and non-inclusive.

We should normalize the possible sex values into another table, rather than using an enum so that they can be customized per LORIS instance according to the study design of the project.

### PR DESCRIPTION
**Pull request by CamilleBeau**
## Brief summary of changes
This PR changed the datatype of the candidate table columns "Sex" and ProbandSex" from enum to varchar(255), instead creating a constraint on the columns determined by a new "sex" table. 

The "sex" table contains optional values for candidate sex and candidate proband sex. Projects can add or remove Sex options by simply inserting into or dropping from this table. 

I also modified anywhere that I could find that was hardcoding Sex options to instead get the sex options from the database through a Utility class function.

- [x] Have you updated related documentation?

#### Testing instructions (if applicable)

1. Test creating a new candidate. Make sure that the sex is inserted into the candidate table properly. 
2. Set "use proband" config setting to true in the config module, and then test the proband tab of candidate parameters
3. Go to the candidate list module and select "Show Advanced Filters". Make sure that the Sex filter contains the right values  and works properly. 
4. Check the cnv, methylation, profiles, and snp tabs all filter for Sex properly, and that the sex dropdown contains the correct values. 
5. Add or drop a value from the Sex table, and repeat steps 1 through 5 making sure that the drop-downs are populated with the updated sex options. 
6. Attempt update a candidate to have a Sex and ProbandSex that do not exist in the sex table. Make sure you get an error.
7. Attempt to delete  a value from the sex table that is used in the candidate table. Make sure you get an error.
8. Test the CouchDB_Import_Demographics.php script

#### Link(s) to related issue(s)

* Resolves #8936

### CODE REVIEW

**Comment by CamilleBeau** [comment]
@kongtiaowang I may need a bit of help with the test suites here. It looks like I'm going to have to add the Sex table to the mock data in these tests, but I don't understand how to do this very well

**Comment by kongtiaowang** [comment]
@CamilleBeau I will debug the test issue later, after https://github.com/aces/Loris/pull/9026 which it will fix the api test issue.

**Comment by CamilleBeau** [comment]
@kongtiaowang Thanks so much!

File: SQL/0000-00-00-schema.sql [hunk 1]
```diff
@@ -72,6 +72,15 @@ CREATE TABLE `language` (
 INSERT INTO language (language_code, language_label) VALUES
     ('en-CA', 'English');
 
+CREATE TABLE `sex` (
+  `SexID` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `SexValue` varchar(255) NOT NULL,
```
Comment (MaximeMulder) [inline SQL/0000-00-00-schema.sql:77]: Nitpicky but as I said during the meeting, I find "SexValue" to be a little generic, and think something like "SexName" would be more descriptive (this also applies to the rest of this PR obviously).

Comment (CamilleBeau) [reply SQL/0000-00-00-schema.sql]: Updated "SexValue" to "Sex", and also modified charset to utf8mb4

**Comment by CamilleBeau** [comment]
@kongtiaowang following the fix for api test issue and rebasing, this should be ready for debugging

File: SQL/0000-00-00-schema.sql [hunk 2]
```diff
@@ -72,6 +72,15 @@ CREATE TABLE `language` (
 INSERT INTO language (language_code, language_label) VALUES
     ('en-CA', 'English');
 
+CREATE TABLE `sex` (
+  `SexID` int(10) unsigned NOT NULL AUTO_INCREMENT,
+  `Sex` varchar(255) NOT NULL,
```
Comment (MaximeMulder) [inline SQL/0000-00-00-schema.sql:77]: Hhhhm, my previous comment was just a little thought I had during our meeting, but if I am to really review this pull request I guess I'll try to be more exhaustive.

After giving it more thought, I am still not fully satisfied with the field names. I think having `Sex` in these names when we are already in the `sex` table is a little redundant. For instance, I'd really prefer to write `sex.ID` and `sex.Name` rather than `sex.SexID` and `sex.Sex` in my SQL requests, especially since the first one may look like a foreign key and the second is quite cryptic (imo). So I strongly advocate for the names `ID` and `Name` for the field names.

Regarding capitalization, the database sure is messy but I think it would be a good idea to maintain a consistent convention in at least the new tables. I am under the impression that this convention is unofficially singular snake_case for the table names and PascalCase for the field names. I am fine with that and this PR matches this convention, but I am still new and this impression of mine may be wrong, so can someone with more database experience than me confirm or deny it ?

Finally, regarding the pointing of sex foreign keys on `ID` or `Name`, I do not have a strong opinion as both have their pros and cons. I think using `ID` as a foreign key would probably be conceptually cleaner, but also require to update the current database and queries as well as require more joins in the latter. Using `Name` is certainly easier and will lead to simpler queries.

It would probably have been easier if I had mentionned all that earlier so sorry for the late reply ! In any case, I'd like someone with more database experience than me to also review this pull request.

**Comment by MaximeMulder** [comment]
Hhhhm, my previous comment was just a little thought I had during our meeting, but if I am to really review this pull request I guess I'll try to be more exhaustive.

After giving it more thought, I am still not fully satisfied with the field names. I think having Sex in these names when we are already in the sex table is a little redundant. For instance, I'd really prefer to write sex.ID and sex.Name rather than sex.SexID and sex.Sex in my SQL requests, especially since the first one may look like a foreign key and the second is quite cryptic (imo). So I strongly advocate for the names ID and Name for the field names.

Regarding capitalization, the database sure is messy but I think it would be a good idea to maintain a consistent convention in at least the new tables. I am under the impression that this convention is unofficially singular snake_case for the table names and PascalCase for the field names. I am fine with that and this PR matches this convention, but I am still new and this impression of mine may be wrong, so can someone with more database experience than me confirm or deny it ?

Finally, regarding the pointing of sex foreign keys on ID or Name, I do not have a strong opinion as both have their pros and cons. I think using ID as a foreign key would probably be conceptually cleaner, but also require to update the current database and queries as well as require more joins in the latter. Using Name is certainly easier and will lead to simpler queries.

It would probably have been easier if I had mentionned all that earlier so sorry for the late reply ! In any case, I'd like someone with more database experience than me to also review this pull request.

File: SQL/0000-00-00-schema.sql (hunk 2 above)
Comment (kongtiaowang) [reply SQL/0000-00-00-schema.sql:77]: ERROR 3780 (HY000) at line 155: Referencing column 'Sex' and referenced column 'Sex' in foreign key constraint 'FK_candidate_sex_1' are incompatible.

Comment (driusan) [reply SQL/0000-00-00-schema.sql:77]: > Regarding capitalization, the database sure is messy but I think it would be a good idea to maintain a consistent convention in at least the new tables. I am under the impression that this convention is unofficially singular snake_case for the table names and PascalCase for the field names. I am fine with that and this PR matches this convention, but I am still new and this impression of mine may be wrong, so can someone with more database experience than me confirm or deny it ?

That seems to match is the official policy documented in `docs/SQLModelingStandard.md` but it's so weakly enforced that I had to go look it up.

File: SQL/0000-00-00-schema.sql [hunk 3]
```diff
@@ -72,6 +72,15 @@ CREATE TABLE `language` (
 INSERT INTO language (language_code, language_label) VALUES
     ('en-CA', 'English');
 
+CREATE TABLE `sex` (
+  `SexID` int(10) unsigned NOT NULL AUTO_INCREMENT,
```
Comment (driusan) [inline SQL/0000-00-00-schema.sql:76]: What purpose does this column have? It is not being used as a foreign key by any other tables and the other column in this table already has a not null/unique constraint and is being used as a foreign key into this table.

Removing it would solve the "what should it be named" debate in the thread below.

Comment (CamilleBeau) [reply SQL/0000-00-00-schema.sql:76]: True, this column is not useful. I don't mind removing it

File: SQL/0000-00-00-schema.sql (hunk 2 above)
Comment (MaximeMulder) [reply SQL/0000-00-00-schema.sql]: Damn I looked at the documentation but somehow missed this file, thanks for pointing it to me. I forgot my glasses at home today so I guess I'm gonna use that as my excuse 😅

File: src/StudyEntities/Candidate/Sex.php [hunk 4]
```diff
@@ -55,13 +53,14 @@ public function __construct(string $value)
...
+     * @param string $value      The value to be validated
+     * @param array  $laidValues The restricted optional values of $value
      *
      * @return bool True if the value format is valid
      */
-    public static function validate(string $value): bool
+    public static function validate(string $value, array $validValues): bool
     {
-        return in_array($value, self::VALID_VALUES, true);
+        return in_array($value, $validValues, true);
```
[the remark this answers is not in the document]
Comment (CamilleBeau) [reply src/StudyEntities/Candidate/Sex.php]: @kongtiaowang Made that change but still having trouble unfortunately!

File: modules/candidate_parameters/php/candidatequeryengine.class.inc [hunk 5]
```diff
@@ -74,7 +74,7 @@ class CandidateQueryEngine extends \LORIS\Data\Query\SQLQueryEngine
                     "Sex",
                     "Candidate's biological sex",
                     $candscope,
-                    new \LORIS\Data\Types\Enumeration('Male', 'Female', 'Other'),
+                    new \LORIS\Data\Types\StringType(255),
```
Comment (driusan) [inline modules/candidate_parameters/php/candidatequeryengine.class.inc:77]: The domain is restricted and we still want the options to be a dropdown. This should be an enum of `\Utility::getSexList()...`, not a string.

File: test/unittests/CandidateTest.php [hunk 6]
```diff
@@ -141,6 +141,11 @@ protected function setUp(): void
         $this->_factory = NDB_Factory::singleton();
         $this->_factory->setConfig($this->_configMock);
         $this->_factory->setDatabase($this->_dbMock);
+        /**
+         * To solve the phan issue
+         *
+         * @phan-methods class \Database
```
Comment (driusan) [inline test/unittests/CandidateTest.php:147]: What is "the phan issue"? Why is this here?

File: test/unittests/CandidateTest.php [hunk 7]
```diff
@@ -157,7 +162,11 @@ protected function setUp(): void
             'RegistrationProjectID' => '1',
             'ProjectTitle'          => '',
         ];
-
+        // Set the expected behavior for the getSexList method
+        /* @phan-suppress-next-line PhanUndeclaredMethod */
```
Comment (driusan) [inline test/unittests/CandidateTest.php:166]: We should not be suppressing errors detected by whan. What is it detecting as undeclared?
